src/prefect/engine/task_runner.py

```diff
@@ -430,12 +430,12 @@ def check_task_trigger(
             - ENDRUN: if the trigger raises an error
         """
 
-        all_states = set()  # type: Set[State]
-        for upstream_state in upstream_states.values():
+        # Deep copy upstream states and expand mapped states
+        all_states = copy.deepcopy(upstream_states)
+        for edge, upstream_state in upstream_states.items():
             if isinstance(upstream_state, Mapped):
-                all_states.update(upstream_state.map_states)
-            else:
-                all_states.add(upstream_state)
+                for mapped_state in upstream_state.map_states:
+                    all_states[copy.deepcopy(edge)] = mapped_state
```

Comment:
Is there a better way to do this? 🤔 I was thinking that we make copies of the edges to put in the dictionary but now I can forsee how these edges will be hard to reason with (e.g. multiple copies of the "same" edge used as keys)

Maybe the values should be a list of States

Reply:
Note: this implementation doesn't work.  You can't have the same key with multiple values -- only the last set value will have any effect here.

Reply:
I'm tempted to say you can let the trigger handle this --> within the trigger it can detect the presence of Mapped upstreams (via the Edge) and do whatever it wants to do (such as unpacking `map_states`).

Reply:
In fact, I think you could take the logic from within the task runner and put it inside each new trigger function and be good.